### DISCUSSION TITLE
abseil logging / glog migration: log-levels

### DISCUSSION THREAD

**Discussion by sschnug** [abseil/abseil-cpp#1343] [General]
Hi,

i'm trying to see a possible migration path of *Glog* to *abseil-cpp*s *logging-module* and are struggling a bit in regards to *log-levels*.

Having heavily relied on `VLOG(some_integer)` in the past, i can't see a way to continue with this approach seeing abseil-cpps log-severities (https://github.com/abseil/abseil-cpp/blob/master/absl/base/log_severity.h) because of the cardinality of predefined levels of the latter.

While having a small set of log-levels is probably a good idea in general and popular among various logging-libraries (e.g. spdlog https://github.com/gabime/spdlog/blob/4f800773393d3ebac13c1fcd946a315d4d72bcd9/include/spdlog/common.h#L211), i only see 3 levels (and a 4th with special semantics).

I'm personally kind of missing `DEBUG` and maybe `TRACE` levels (if we stay in predefined-level world). 
*abseil-py* surprisingly has support for `DEBUG` (https://abseil.io/docs/python/guides/logging).

Having well-supported access to *integer-based log-levels* would even be better. 

*or-tools*, somewhat reflecting my use-cases (library-code with easy to brake combinatorial-stuff in very deep hierarchies) uses 6 different log-levels too (according to a quick look: https://github.com/google/or-tools/search?q=vlog , link seems to automatically die, but it's a simple search for `VLOG`).

Maybe this has something to do with abseil-cpp focusing on service-code and less on library-code?

**Is there any oversight on my end or are there any tips on achieving a more fine-granular log-control?**

(it seems one could introduce negative levels > kNegativeInfinity, but i'm not sure about the effects of `NormalizeLogSeverity` and it also feels wrong)

Thanks for reading!

**Comment by derekmauro** [comment]
We are working on releasing `VLOG`. Please be patient.

**Reply by nlohmann** [reply]
Now that the first version of the Abseil logging library is out there, is there a plan when to expect `VLOG` or more log levels?

**Reply by jiawen** [reply]
I also noticed that the first release candidate for the LTS is available. It would be amazing if VLOG were available as part of the LTS. It would declare absl logging "feature complete" enough for long term support.

**Reply by derekmauro** [reply]
VLOG won't make the Jan 2023 LTS. Sorry. We are still working on it.

**Reply by jiawen** [reply]
No worries and thanks for the heads up! Keep abseil amazing!

**Reply by nlohmann** [reply]
@derekmauro Any update on this?

**Reply by derekmauro** [reply]
I actually had a discussion with the person working on it last week. It was stalled for a while but it seems like progress is being made again. Unfortunately I still don't have a good estimate for when you can expect it.

**Reply by nlohmann** [reply]
Thanks for the update!

**Reply by sfc-gh-ebrossard** [reply]
I'm a little sad that VLOG didn't make it into the August LTS release. I take it that there's still no ETA, right?

**Reply by jiawen** [reply]
I'm sad too. VLOG is the only reason my team's monorepo is still on glog.

**Reply by derekmauro** [reply]
I get that and I'm not happy about it either. We did just release `DFATAL` after the release was cut. There is still no ETA, but we are a bit closer.

**Reply by jiawen** [reply]
No pressure! I'm happy to help in any way I can (testing on funny platforms, etc).

**Reply by derekmauro** [reply]
I got so frustrated with the lack of progress on this that I went and pushed it though myself. Please test out https://github.com/abseil/abseil-cpp/commit/7b6c17e378224844d8663a410da2da5353b205b9 and let us know if you encounter any issues.

**Reply by sfc-gh-ebrossard** [reply]
Awesome, thanks so much! It's like an early Christmas present 😀 I'll try to plug in Abseil's logging soon for our team's current glog usage and see how it goes.

By the way, I'm not sure how much it matters, but there's a Buganizer link and a reference to TAP in the commit above.

**Reply by derekmauro** [reply]
No problem!

With regards to the buganizer link: At one point we redacted the buganizer links but we are now ok publishing them. Differences between our internal repo and what makes it to GitHub just make it harder to sync code.

**Comment by sschnug** [comment]
I guess this can be closed now as the feature is now available. Thanks!
